### PR TITLE
fix: replace gundb with kv storage

### PR DESCRIPTION
**Pull request by lelenei**
## Description

Closes # (NO_ISSUE)

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/web3-providers/src/index.ts [hunk 1]
```diff
@@ -6,6 +6,8 @@ import { NativeExplorerAPI } from './explorer'
 import { RSS3API } from './rss3'
 import { KeyValueAPI } from './kv'
 
+export * from './kv'
```
Comment (guanbinrui) [inline packages/web3-providers/src/index.ts]:
```suggestion
```

File: packages/web3-providers/src/types.ts [hunk 2]
```diff
@@ -291,7 +291,7 @@ export namespace NonFungibleTokenAPI {
 export namespace StorageAPI {
     export interface Storage {
         set<T extends {}>(key: string, value: T): Promise<void>
-        get<T>(key: string): Promise<T | void>
+        get<T>(key: string): Promise<T | undefined>
         delete?: (key: string) => Promise<void>
```
Comment (septs) [inline packages/web3-providers/src/types.ts]:
```suggestion
        delete?(key: string): Promise<void>
```

Comment (guanbinrui) [reply packages/web3-providers/src/types.ts]: Resolve this here https://github.com/DimensionDev/Maskbook/pull/5462.

File: packages/mask/src/plugins/Avatar/Services/gun.ts [hunk 3]
```diff
@@ -1,54 +1,43 @@
...
-import { gun2 } from '../../../network/gun/version.2'
+import { ChainId, isSameAddress } from '@masknet/web3-shared-evm'
 import { NFT_AVATAR_GUN_SERVER } from '../constants'
+import { NetworkPluginID } from '@masknet/plugin-infra'
+import addSeconds from 'date-fns/addSeconds'
+import isBefore from 'date-fns/isBefore'
+import { KeyValue } from '@masknet/web3-providers'
 
-const NFTAvatarGUN = gun2.get(NFT_AVATAR_GUN_SERVER)
+const NFTAvatarDB = KeyValue.createJSON_Storage(NFT_AVATAR_GUN_SERVER)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Avatar/Services/gun.ts]:
```suggestion
const NFTAvatarDB = KeyValue.createJSON_Storage(NFT_AVATAR_DB_NAME)
```

File: packages/mask/src/plugins/Avatar/Services/gun.ts [hunk 4]
```diff
@@ -1,54 +1,43 @@
...
-            //@ts-expect-error
-            .get(userId).then!()
-    }
+async function _getUserAddress(userId: string, networkPluginId?: NetworkPluginID, chainId?: ChainId) {
+    const result = await NFTAvatarDB.get<Record<string, string>>(userId)
+    return result?.[`${networkPluginId ?? NetworkPluginID.PLUGIN_EVM}-${chainId ?? ChainId.Mainnet}`]
+}
 
-    return result
+export async function getUserAddress(userId: string, networkPluginId?: NetworkPluginID, chainId?: ChainId) {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Avatar/Services/gun.ts]:
```suggestion
export async function getUserAddress(userId: string, networkPluginId?: NetworkPluginID, chainId?: number) {
```

File: packages/mask/src/plugins/Avatar/Services/gun.ts [hunk 5]
```diff
@@ -1,54 +1,43 @@
...
+    c = cache.get(userId)
+    return c?.[0]
 }
 
-export async function setUserAddress(userId: string, address: string) {
+export async function setUserAddress(
+    userId: string,
+    address: string,
+    networkPluginId?: NetworkPluginID,
+    chainId?: ChainId,
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Avatar/Services/gun.ts]:
```suggestion
    chainId?: number,
```

File: packages/mask/src/plugins/Avatar/Services/bind.ts [hunk 6]
```diff
@@ -9,8 +9,12 @@ const NFTAvatarDB = KeyValue.createJSON_Storage(NFT_AVATAR_DB_NAME)
 const cache = new Map<string, [Promise<string | undefined>, number]>()
 
 async function _getUserAddress(userId: string, networkPluginId?: NetworkPluginID, chainId?: number) {
-    const result = await NFTAvatarDB.get<Record<string, string>>(userId)
-    return result?.[`${networkPluginId ?? NetworkPluginID.PLUGIN_EVM}-${chainId ?? ChainId.Mainnet}`]
+    try {
+        const result = await NFTAvatarDB.get<Record<string, string>>(userId)
+        return result?.[`${networkPluginId ?? NetworkPluginID.PLUGIN_EVM}-${chainId ?? ChainId.Mainnet}`]
```
Comment (septs) [inline packages/mask/src/plugins/Avatar/Services/bind.ts]: plz extract this template-literal expression as function

getter and setter need consistent key generate algorithm

Comment (septs) [reply packages/mask/src/plugins/Avatar/Services/bind.ts]: e.g: `getKey(networkPluginId, chainId)`

Comment (septs) [reply packages/mask/src/plugins/Avatar/Services/bind.ts]: @lelenei @guanbinrui

File: packages/mask/src/plugins/Avatar/Services/bind.ts [hunk 7]
```diff
@@ -3,18 +3,37 @@ import { NetworkPluginID } from '@masknet/plugin-infra'
...
+    return
+}
+
 async function _getUserAddress(userId: string, networkPluginId?: NetworkPluginID, chainId?: number) {
     try {
         const result = await NFTAvatarDB.get<{ networkPluginId: string; chainId: number; address: string }>(userId)
         if (!result || !result?.address) {
             const result = await NFTAvatarDBStorage.get<Record<string, string>>(userId)
-            return result?.[`${networkPluginId ?? NetworkPluginID.PLUGIN_EVM}-${chainId ?? ChainId.Mainnet}`]
+            const address = result?.[`${networkPluginId ?? NetworkPluginID.PLUGIN_EVM}-${chainId ?? ChainId.Mainnet}`]
```
Comment (septs) [inline packages/mask/src/plugins/Avatar/Services/bind.ts]: plz extract the key generate expression as function

File: packages/mask/src/plugins/Avatar/Services/index.ts [hunk 8]
```diff
@@ -1,39 +1,38 @@
...
-    return avatar
+export async function saveNFTAvatar(
+    address: string,
+    nft: AvatarMetaDB,
+    networkPluginId?: NetworkPluginID,
+    chainId?: number,
+) {
+    try {
+        const avatar = await saveNFTAvatarToRSS(address, nft, '')
+        setUserAddress(nft.userId, address, networkPluginId, chainId)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Avatar/Services/index.ts]:
```suggestion
        await setUserAddress(nft.userId, address, networkPluginId, chainId)
```

File: packages/mask/src/social-network-adaptor/twitter.com/injection/NFT/NFTAvatarInTwitter.tsx [hunk 9]
```diff
@@ -89,7 +96,19 @@ function NFTAvatarInTwitter() {
         const avatar = await PluginNFTAvatarRPC.saveNFTAvatar(wallet.address, {
             ...NFTEvent,
             avatarId: getAvatarId(identity.avatar ?? ''),
-        } as AvatarMetaDB)
+        } as AvatarMetaDB).catch((error) => {
+            console.log(error.message)
```
Comment (guanbinrui) [inline packages/mask/src/social-network-adaptor/twitter.com/injection/NFT/NFTAvatarInTwitter.tsx]:
```suggestion
```

Comment (septs) [reply packages/mask/src/social-network-adaptor/twitter.com/injection/NFT/NFTAvatarInTwitter.tsx]: enable no-console rule?
see https://eslint.org/docs/rules/no-console

File: packages/mask/src/plugins/Avatar/Services/bind.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+            .get(userId).then!()
+        if (result) return result
+        await delay(500)
+    }
+    return
+}
+
+function getKey(networkPluginId?: NetworkPluginID, chainId?: number) {
+    return `${networkPluginId ?? NetworkPluginID.PLUGIN_EVM}-${chainId ?? ChainId.Mainnet}`
+}
```
Comment (septs) [inline packages/mask/src/plugins/Avatar/Services/bind.ts]:
```suggestion
function getKey(pluginId = NetworkPluginID.PLUGIN_EVM, chainId: number = ChainId.Mainnet) {
    return `${pluginId}-${chainId}`
}
```

File: packages/mask/src/plugins/Avatar/Services/bind.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    return
+}
+
+function getKey(networkPluginId = NetworkPluginID.PLUGIN_EVM, chainId: number = ChainId.Mainnet) {
+    return `${networkPluginId}-${chainId}`
+}
+
+function getCacheKey(userId: string, networkPluginId = NetworkPluginID.PLUGIN_EVM, chainId: number = ChainId.Mainnet) {
+    return `${userId}-${networkPluginId}-${chainId}`
+}
```
Comment (septs) [inline packages/mask/src/plugins/Avatar/Services/bind.ts:35]: why?

File: packages/web3-providers/src/helpers.ts [hunk 12]
```diff
@@ -1,5 +1,4 @@
 export async function fetchJSON<T = unknown>(requestInfo: RequestInfo, requestInit?: RequestInit): Promise<T> {
     const res = await globalThis.fetch(requestInfo, requestInit)
-    if (res.status === 404) return {} as T
```
Comment (septs) [inline packages/web3-providers/src/helpers.ts]: CC @guanbinrui

**Review by guanbinrui — DISMISSED** [review]

**Pull request merged**